pkg/services/ngalert/api.go

```diff
@@ -0,0 +1,183 @@
+		"instances": instances,
+	})
+}
+
+// alertDefinitionEval handles GET /api/alert-definitions/eval/:dashboardId/:panelId/:refId".
+func (ng *AlertNG) alertDefinitionEval(c *models.ReqContext) api.Response {
+	alertDefinitionID := c.ParamsInt64(":alertDefinitionId")
+
+	fromStr := c.Query("from")
+	if fromStr == "" {
+		fromStr = "now-3h"
+	}
+
+	toStr := c.Query("to")
+	if toStr == "" {
+		toStr = "now"
+	}
+
+	conditions, err := ng.LoadAlertCondition(alertDefinitionID, c.SignedInUser, c.SkipCache)
+	if err != nil {
+		return api.Error(400, "Failed to load conditions", err)
+	}
+
+	alertCtx, cancelFn := context.WithTimeout(context.Background(), setting.AlertingEvaluationTimeout)
+	defer cancelFn()
+
+	alertExecCtx := eval.AlertExecCtx{Ctx: alertCtx, SignedInUser: c.SignedInUser}
+
+	execResult, err := conditions.Execute(alertExecCtx, fromStr, toStr)
+	if err != nil {
+		return api.Error(400, "Failed to execute conditions", err)
+	}
+
+	evalResults, err := eval.EvaluateExecutionResult(execResult)
+	if err != nil {
+		return api.Error(400, "Failed to evaluate results", err)
+	}
+
+	frame := evalResults.AsDataFrame()
+
+	df := tsdb.NewDecodedDataFrames([]*data.Frame{&frame})
+	instances, err := df.Encoded()
+	if err != nil {
+		return api.Error(400, "Failed to encode result dataframes", err)
+	}
+
+	return api.JSON(200, util.DynMap{
+		"instances": instances,
+	})
+}
+
+// getAlertDefinitionEndpoint handles GET /api/alert-definitions/:alertDefinitionId.
+func (ng *AlertNG) getAlertDefinitionEndpoint(c *models.ReqContext) api.Response {
+	alertDefinitionID := c.ParamsInt64(":alertDefinitionId")
+
+	query := getAlertDefinitionByIDQuery{
+		ID: alertDefinitionID,
+	}
+
+	if err := ng.getAlertDefinitionByID(&query); err != nil {
+		return api.Error(500, "Failed to get alert definition", err)
+	}
+
+	return api.JSON(200, &query.Result)
+}
+
+// deleteAlertDefinitionEndpoint handles DELETE /api/alert-definitions/:alertDefinitionId.
+func (ng *AlertNG) deleteAlertDefinitionEndpoint(c *models.ReqContext) api.Response {
+	alertDefinitionID := c.ParamsInt64(":alertDefinitionId")
+
+	query := deleteAlertDefinitionByIDQuery{
+		ID:    alertDefinitionID,
+		OrgID: c.SignedInUser.OrgId,
+	}
+
+	if err := ng.deleteAlertDefinitionByID(&query); err != nil {
+		return api.Error(500, "Failed to delete alert definition", err)
+	}
+
+	return api.JSON(200, util.DynMap{"affectedRows": query.RowsAffected})
+}
+
+// updateAlertDefinitionEndpoint handles PUT /api/alert-definitions/:alertDefinitionId.
+func (ng *AlertNG) updateAlertDefinitionEndpoint(c *models.ReqContext, cmd updateAlertDefinitionCommand) api.Response {
+	cmd.ID = c.ParamsInt64(":alertDefinitionId")
+	cmd.SignedInUser = c.SignedInUser
+	cmd.SkipCache = c.SkipCache
+
+	if err := ng.updateAlertDefinition(&cmd); err != nil {
+		return api.Error(500, "Failed to update alert definition", err)
+	}
+
+	return api.JSON(200, util.DynMap{"affectedRows": cmd.RowsAffected, "id": cmd.Result.Id})
```

Comment:
It seems for me when updating an alert the Id is coming back as zero (for an alert that is not id 0).